pipelines.md

```diff
@@ -0,0 +1,358 @@
+
+Example: `foo = ... |> bar() as baz |> qux()`, here the final result is assigned into `foo` (no change to semantics), the result of executing `|> bar()` is assigned into `baz` (equivalent to `|> bar($baz)` in the current syntax).
+
+The earlier example becomes:
+
+```
+sketch006 = startSketchOn('XZ')
+  |> startProfileAt([0.1, 1])
+  |> line([0.1, 0])
+  |> angledLineToX({ angle: 10, to: 0.05 })
+  |> yLine(10)
+  |> line([0.6, 0])
+  |> yLine(-.05)
+  |> tangentialArc({ radius: 0.6, offset: -90 }) as arcSketch
+  |> lineTo([profileStartX(arcSketch), profileStartY(arcSketch)])
+  |> close()
+  |> revolve({ axis: 'y' })
+```
+
+### Alternative
+
+Rather than making `as` part of the pipeline syntax, it could be allowed in any expression. The above description would still apply, but it could also be applied to sub-expressions, e.g., `hole(circle(...) as innerCircle)`.
+
+This would be more flexible and would avoid learners hitting the issue of knowing where `as` can be used ("I can use it here and here, so why can't I use it here?"). However, I believe it would lead to poor programming style: it is generally better to use a variable declaration to refer to reused geometry or data rather than use `as` since it will be easier to scan code for declarations and it encourages a less nested, more straightforward coding style. On the other hand, pipelining is ergonomic and allowing `as` there prevents breaking pipelines arbitrarily.
+
+The precedence in general sub-expressions may be confusing, e.g., in `a + b as c + d`, does `c` refer to the value of `b` or of `a + b`? 
+
+## Braced blocks in pipelines
+
+Braced blocks are allowed in pipelines. The current receiver of the pipeline is piped into the *last* expression of the block (the block must have a last expression, similar to `if` blocks). The result of the block is the result of the final expression. Variables within the block are scoped to the block. (Possible extension: use `export x = ...` or `... export as x` to make variables visible in the enclosing scope).
+
+```
+startSketchOn('XZ')
+  |> {
+    x = sin(42)
+    y = cos(42)
+    line([x, y])
+  }
+  |> close()
+
+// Or equivalently
+
+startSketchOn('XZ')
+  |> {
+    x = sin(42)
+    y = cos(42)
+    line([x, y])
+      |> close()
+  }
+```
+
+This allows more precise scoping of variables (in the above example, `x` and `y` are not available outside the block), but the primary motivator is for use in `for` expressions, see below.
+
+## `for` expressions
+
+Syntax: `'for' expr_1 'as' id |> expr_2`. May appear within a pipeline or as an expression outside a pipeline.
+
+In the simplest case (with no preceding pipeline), `expr_1` is evaluated to some kind of an object `s` with sequence type (an array or range). `expr_2` is evaluated with each item in `s` bound in turn to `id`. The result of execution is an array of the results of evaluating `expr_2` (i.e., you can think of the expression as performing 'for each' or 'map' on its input).
```

Comment:
Bear in mind this description is intended for implementers rather than users. I would start with the examples and ask what they suggest? Also maybe better to start with the `for i in [...]` version rather than the `as` version.